quickwit/quickwit-control-plane/src/ingest/ingest_controller.rs

```diff
@@ -444,67 +545,45 @@ impl IngestController {
             );
             return None;
         }
-        let mut leader_follower_pairs = Vec::with_capacity(num_shards_to_allocate);
 
-        let mut num_open_shards: usize = 0;
-        let mut per_leader_num_open_shards: HashMap<&str, usize> =
-            HashMap::with_capacity(num_ingesters);
+        // Count of open shards per available ingester node (including the ingester with 0 open
+        // shards).
+        let mut per_node_num_open_shards: HashMap<&str, usize> = ingesters
+            .iter()
+            .map(|ingester| (ingester.as_str(), 0))
+            .collect();
 
         for shard in model.all_shards() {
             if shard.is_open() && !unavailable_leaders.contains(&shard.leader_id) {
-                num_open_shards += 1;
-
-                *per_leader_num_open_shards
-                    .entry(&shard.leader_id)
-                    .or_default() += 1;
-            }
-        }
-        let mut num_remaining_shards_to_allocate = num_shards_to_allocate;
-        let num_open_shards_target = num_shards_to_allocate + num_open_shards;
-        let max_num_shards_to_allocate_per_node = num_open_shards_target / num_ingesters;
-
-        // Allocate at most `max_num_shards_to_allocate_per_node` shards to each ingester.
-        for (leader_id, follower_id) in ingesters.iter().zip(ingesters.iter().cycle().skip(1)) {
-            if num_remaining_shards_to_allocate == 0 {
-                break;
-            }
-            let num_open_shards_inner = per_leader_num_open_shards
-                .get(leader_id.as_str())
-                .copied()
-                .unwrap_or_default();
-
-            let num_shards_to_allocate_inner = max_num_shards_to_allocate_per_node
-                .saturating_sub(num_open_shards_inner)
-                .min(num_remaining_shards_to_allocate);
-
-            for _ in 0..num_shards_to_allocate_inner {
-                num_remaining_shards_to_allocate -= 1;
-
-                let leader = leader_id.clone();
-                let mut follower_opt = None;
-
-                if self.replication_factor > 1 {
-                    follower_opt = Some(follower_id.clone());
+                if let Some(shard_count) =
+                    per_node_num_open_shards.get_mut(shard.leader_id.as_str())
+                {
+                    *shard_count += 1;
+                }
+                if let Some(follower_id) = shard.follower_id.as_ref() {
+                    if let Some(shard_count) =
+                        per_node_num_open_shards.get_mut(follower_id.as_str())
+                    {
+                        *shard_count += 1;
+                    }
                 }
```

Comment:
is it possible to have a shard not present in `per_node_num_open_shards`?
Should we log something when it happens?

Reply:
It is normal if it is not there, and it does not require a log. I added a comment about that.